File: src/main/java/org/fountainmc/api/world/block/Slab.java
package org.fountainmc.api.world.block;

import javax.annotation.Nonnull;

@Nonnull
public interface Slab extends BlockState {

    /**
     * Get the block type of the slab
     *
     * @return the type of the slab
     */
    StairType getType();

    /**
     * Return a copy of this slab with the given slab typez
     *
     * @param type the new type
     * @return the new slab
     */
    Slab withType(StairType type);

    /**
     * If the slab is upside-down, and occupies the top half of its tile.
     *
     * @return if upside-down
     */
    boolean isUpsideDown();

    /**
     * If the slab is rightside-up, and occupies the bottom half of its tile.
     *
     * @return if rightside-up
     */
    boolean isRightsideUp();

    StairBlock withUpsideDown(boolean b);

    default StairBlock asRightsideUp() {
        return withUpsideDown(false);
    }

    default StairBlock asUpsideDown() {
        return withUpsideDown(false);
    }

}

File: src/main/kotlin/org/fountainmc/api/world/block/Slab.kt
package org.fountainmc.api.world.block

interface Slab : BlockState {

    /**
     * Get the block type of the slab

     * @return the type of the slab
     */
    val type: StairType

    /**
     * Return a copy of this slab with the given slab typez

     * @param type the new type
     * *
     * @return the new slab
     */
    fun withType(type: StairType): Slab

    /**
     * If the slab is upside-down, and occupies the top half of its tile.

     * @return if upside-down
     */
    val isUpsideDown: Boolean

    /**
     * If the slab is rightside-up, and occupies the bottom half of its tile.

     * @return if rightside-up
     */
    val isRightsideUp: Boolean

    fun withUpsideDown(b: Boolean): StairBlock

    fun asRightsideUp(): StairBlock {
        return withUpsideDown(false)
    }

    fun asUpsideDown(): StairBlock {
        return withUpsideDown(false)
    }

}
